pep-9999.rst

```diff
@@ -0,0 +1,213 @@
+PEP: 9999
+Title: Inline variance operators for generic type variables: ``+T`` and ``-T``
+Author: Joren Hammudoglu <jhammudoglu at gmail.com>
+Status: Active
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 24-Jul-2021
+Post-History: 24-Jul-2021
+
+
+Abstract
+========
+
+PEP 484 proposes a syntax for defining the `covariance and contravariance 
+<https://www.python.org/dev/peps/pep-0484/#covariance-and-contravariance>`_
+of `user-defined generic types
+<https://www.python.org/dev/peps/pep-0484/#user-defined-generic-types>`_::
+
+    T_co = typing.TypeVar('T_co', covariant=True)
+    T_contra = typing.TypeVar('T_contra', contravariant=True)
+
+
+    # a covariant generic type
+    class EventDispatcher(typing.Generic[T_co]):
+        def dispatch() -> T_co: ...
+
+    # a contravariant generic type
+    class EventListener(typing.Generic[T_contra]):
+        def receive(event: T_contra) -> None: ...
+
+
+This PEP proposes a concise syntax for specifying the variance of
+generic types by overloading the ``+`` and ``-`` prefix operators
+on type variables::
+
+    T = typing.TypeVar('T')
+
+
+    # a covariant generic type
+    class EventDispatcher(typing.Generic[+T]):
+        def dispatch() -> T: ...
+
+    # a contravariant generic type
+    class EventListener(typing.Generic[-T]):
+        def receive(event: T) -> None: ...
+
+
+
+Rationale
+=========
+
+Although defining the variance of generic type by using type variables
+that are constructed with the ``covariant=True`` or ``contravariant=True``
+keyword arguments, it has downsides. 
+
+This is why type checkers do no allow using co- or contravariant type
+variables outside of generic classes. 
+
+
+- In PEP 484 it states:
+
+    Covariance or contravariance is not a property of a type variable,
+    but a property of a generic class defined using this variable. 
+
+  Instead of defining variance on the type variable itself, it would
+  make more sense to mark the type variable as variant when defining
+  the generic type, and only there. 
+
+- Because variance is a property of the generic type itself, type 
+  checkers do no allow using co- or contravariant type variables 
+  outside of generic classes. But as the variance is currently a
+  property of the type variables, the must be used within variant
+  generic classes.
+
+- Creating separate type variables for different variances often
```

Comment:
Not sure this is really common. It would be useful to have examples from open source projects where you see this duplication.